app/controllers/supplementary/PackageInformationController.scala

```diff
@@ -31,31 +35,85 @@ import views.html.supplementary.package_information
 import scala.concurrent.{ExecutionContext, Future}
 
 class PackageInformationController @Inject()(
-  appConfig: AppConfig,
-  override val messagesApi: MessagesApi,
   authenticate: AuthAction,
-  customsCacheService: CustomsCacheService
-)(implicit ec: ExecutionContext)
+  errorHandler: ErrorHandler,
+  cacheService: CustomsCacheService
+)(implicit ec: ExecutionContext, appConfig: AppConfig, override val messagesApi: MessagesApi)
     extends FrontendController with I18nSupport {
-  import forms.supplementary.PackageInformation._
+
+  val packagesMaxElements = 99
 
   def displayForm(): Action[AnyContent] = authenticate.async { implicit request =>
-    customsCacheService.fetchAndGetEntry[PackageInformation](supplementaryCacheId, formId).map {
-      case Some(data) => Ok(package_information(appConfig, form.fill(data)))
-      case _          => Ok(package_information(appConfig, form))
+    cacheService
+      .fetchAndGetEntry[Seq[PackageInformation]](supplementaryCacheId, formId)
+      .map(items => Ok(package_information(form, items.getOrElse(Seq.empty))))
+
+  }
+
+  def submitForm(): Action[AnyContent] = authenticate.async { implicit authRequest =>
+    val actionTypeOpt = authRequest.body.asFormUrlEncoded.map(FormAction.fromUrlEncoded(_))
+    cacheService
+      .fetchAndGetEntry[Seq[PackageInformation]](supplementaryCacheId, formId)
+      .flatMap { data =>
+        implicit val packagings = data.getOrElse(Seq.empty)
+
+        actionTypeOpt match {
+          case Some(Add)             => addItem()
+          case Some(Remove(ids))     => remove(ids.headOption)
+          case Some(SaveAndContinue) => continue()
+          case _                     => errorHandler.displayErrorPage()
+        }
+      }
+  }
+
+  def remove(
+    id: Option[String]
+  )(implicit authRequest: AuthenticatedRequest[AnyContent], packagings: Seq[PackageInformation]) =
+    id match {
+      case Some(id) => {
+        val updatedPackagings =
+          packagings.zipWithIndex.filterNot(_._2.toString == id).map(_._1)
+
+        cacheService
+          .cache[Seq[PackageInformation]](supplementaryCacheId(), formId, updatedPackagings)
+          .map(_ => Redirect(routes.PackageInformationController.displayForm()))
+      }
+      case _ => errorHandler.displayErrorPage()
     }
+
+  def continue()(implicit request: AuthenticatedRequest[AnyContent], packagings: Seq[PackageInformation]) = {
+    val payload = form.bindFromRequest()
+    if (payload.data.filter(_._2.size > 0).size > 1) badRequest(payload, USE_ADD)
```

Comment:
this is the approach i would want every multiple items screen to adhere to.

Reply:
So @raghu1978 it'll be a problem, we decided to stay with the current implementation: `Save and continue` we had few meetings about it. I agree with Paulo, I want to have Save and Continue here.

Based on the implementation instead of using `_._2` `> 1` where you can't know from this code what do you want to filter etc to extract this logic.

```
private def isFormEmpty[A](form: Form[A]): Boolean =
    retrieveData(form).filter { case (_, value) => value.nonEmpty }.isEmpty

  private def retrieveData[A](form: Form[A]): Map[String, String] =
    form.data.filter { case (name, _) => name != "csrfToken" }
```

First question why `1`? For me it's not readable

Reply:
ok will change this